lib/ddtrace/contrib/grape/endpoint.rb

```diff
@@ -0,0 +1,156 @@
+require 'ddtrace/ext/http'
+require 'ddtrace/ext/errors'
+
+module Datadog
+  module Contrib
+    module Grape
+      # Endpoint module includes a list of subscribers to create
+      # traces when a Grape endpoint is hit
+      module Endpoint
+        KEY_RUN = 'datadog_grape_endpoint_run'.freeze
+        KEY_RENDER = 'datadog_grape_endpoint_render'.freeze
+
+        def self.subscribe
+          # Grape is instrumented only if it's available
+          return unless defined?(::Grape) && defined?(::ActiveSupport::Notifications)
+
+          # subscribe when a Grape endpoint is hit
+          ::ActiveSupport::Notifications.subscribe('endpoint_run.grape.start_process') do |*args|
+            endpoint_start_process(*args)
+          end
+          ::ActiveSupport::Notifications.subscribe('endpoint_run.grape') do |*args|
+            endpoint_run(*args)
+          end
+          ::ActiveSupport::Notifications.subscribe('endpoint_render.grape.start_render') do |*args|
+            endpoint_start_render(*args)
+          end
+          ::ActiveSupport::Notifications.subscribe('endpoint_render.grape') do |*args|
+            endpoint_render(*args)
+          end
+          ::ActiveSupport::Notifications.subscribe('endpoint_run_filters.grape') do |*args|
+            endpoint_run_filters(*args)
+          end
+        end
+
+        def self.endpoint_start_process(*)
+          return if Thread.current[KEY_RUN]
+
+          # retrieve the tracer from the PIN object
+          pin = Datadog::Pin.get_from(::Grape)
```

Comment:
Yup, not worth the complexity of attaching the PIN to instances, I'm just saying -> let's keep aware we have this limitation.